gtsfm/densify/interface_patchmatchnet.py

```diff
@@ -0,0 +1,209 @@
+"""Interface class from gtsfmData to patchmatchnetData
+
+Authors: Ren Liu
+"""
+from typing import Any, Dict, List, Tuple
+
+import cv2
+import numpy as np
+from torch.utils.data import Dataset
+
+from gtsfm.common.image import Image
+from gtsfm.common.gtsfm_data import GtsfmData
+from gtsfm.densify.mvs_math import piecewise_gaussian, to_camera_coordinates
+
+
+class PatchmatchNetData(Dataset):
+    """PatchmatchNetData class for Patchmatch Net. It contains the interface from GtsfmData"""
+
+    def __init__(self, images: Dict[int, Image], sfm_result: GtsfmData, num_views: int = 5) -> None:
+        """Initialize method for PatchmatchnetData
+
+        Args:
+            images: input images (H, W, C) to GTSFM
+            sfm_result: sfm results calculated by GTSFM
+            num_views: number of views, containing 1 reference view and (num_views-1) source views
+        """
+
+        assert images is not None and len(images) > 1
+
+        # cache sfm result
+        self.sfm_result = sfm_result
+
+        # Patchmatch Net meta
+        self.num_views = num_views
+        self.num_stages = 4
+
+        # Test data preparation
+        self.keys = sorted(self.sfm_result.get_valid_camera_indices())
+        self.num_images = len(self.keys)
+        self.keys_map = {}
+        for i in range(self.num_images):
+            self.keys_map[self.keys[i]] = i
+        self.images = images
+        self.image_w = images[self.keys[0]].width
+        self.image_h = images[self.keys[0]].height
+
+        self.pairs, self.depth_metas = self.configure()
+
+    def configure(self) -> Tuple[np.ndarray, np.ndarray]:
+        """Configure pairs and depth_metas for each view from sfm_result
+
+        Returns:
+            pairs: np.ndarray in shape of (num_images, num_views-1). Each row_id indicates the index of reference view
+                in self.keys, with (num_views-1) values indicating the indices of source views in self.keys
+            depth_metas: np.ndarray in shape of (num_images, 2). Each row_id indicates the index of reference view
+                in self.keys, with 2 values indicating [min_depth, max_depth]
+        """
+        num_images = self.num_images
+        num_tracks = self.sfm_result.number_tracks()
+
+        pair_scores = np.zeros((num_images, num_images))
+        for i in range(num_images):
+            pair_scores[i, i] = -np.inf
+
+        depth_metas = np.zeros((num_images, 2))
+        depth_collection_views: List[List[float]] = [[] for _ in range(num_images)]
+
+        for i in range(num_tracks):
```

Comment:
inline comment here would help explain why the double for loop is necessary/ what algo is being run here